gettsim/benefits/alg2.py

```diff
@@ -221,58 +221,47 @@ def grossinc_alg2(household):
     )
 
 
-def einkommensanrechnungsfrei(household, params):
-    """Determine the amount of income that is not deducted. Varies withdrawal rates
-    depending on monthly earnings."""
+def e_anr_frei(household, params):
+    """Determine the amount of income that is not deducted. Withdrawal rates
+    depend on monthly earnings and on the number of kids in the household.
+    § 30 SGB II. Since 01.04.2011 § 11b """
     # Calculate the amount of children below the age of 18.
-    child0_18_num = (household["child"] & household["age"].between(0, 18)).sum()
-
-    # 100€ is always 'free'
-    household.loc[(household["m_wage"] <= params["a2grf"]), "ekanrefrei"] = household[
-        "m_wage"
-    ]
-    # until 1000€, you may keep 20% (withdrawal rate: 80%)
-    household.loc[
-        (household["m_wage"].between(params["a2grf"], params["a2eg1"])), "ekanrefrei",
-    ] = params["a2grf"] + params["a2an1"] * (household["m_wage"] - params["a2grf"])
-    # from 1000 to 1200 €, you may keep only 10%
-    household.loc[
-        (household["m_wage"].between(params["a2eg1"], params["a2eg2"]))
-        & (child0_18_num == 0),
-        "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (household["m_wage"] - params["a2eg1"])
-    )
-    # If you have kids, this range goes until 1500 €,
-    household.loc[
-        (household["m_wage"].between(params["a2eg1"], params["a2eg3"]))
-        & (child0_18_num > 0),
-        "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (household["m_wage"] - params["a2eg1"])
-    )
-    # beyond 1200/1500€, you can't keep anything.
-    household.loc[
-        (household["m_wage"] > params["a2eg2"]) & (child0_18_num == 0), "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (params["a2eg2"] - params["a2eg1"])
-    )
-    household.loc[
-        (household["m_wage"] > params["a2eg3"]) & (child0_18_num > 0), "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (params["a2eg3"] - params["a2eg1"])
-    )
-    # Children income is fully deducted, except for the first 100 €.
-    household.loc[(household["child"]), "ekanrefrei"] = np.maximum(
-        0, household["m_wage"] - 100
+    num_childs_0_18 = (household["child"] & household["age"].between(0, 18)).sum()
+
+    if num_childs_0_18 == 0:
+        top_limit_2nd_interval = params["a2eg2"]
+    else:
+        top_limit_2nd_interval = params["a2eg3"]
+
+    cols = ["m_wage", "ekanrefrei"]
+    household.loc[:, cols] = household.groupby("pid")[cols].apply(
```

Comment:
Does that construct work when `household` is a slice of a bigger df? I.e., does it fall back on in-place assignment instead of creating or replacing a column?

Reply:
Then, household is probably a view and it works but raises the infamous ``SettingWithCopyWarning``. The warning is independent of whether you use ``.loc[:`` or not

Reply:
Both raises at least in the test no such warning.

Reply:
Yes, it is not clear whether ``groupby`` groups are copies or views but https://stackoverflow.com/questions/52711823 indicates they are most likely copies. Thus, no warning here.

My point was more general if you do not use groupby.

Reply:
Are you sure, @tobiasraabe? My own model of how things work is different, but I am happy to adjust that! 

@MaxBlesch: Is this being called from a test that uses groupby apply on a bigger DataFrame? And did we not have code somewhere shutting warnings off?

Reply:
It is called from `tax_transfer` test and there we have a larger dataset. This warning is not turned off. See https://github.com/iza-institute-of-labor-economics/gettsim/blob/issue110/tox.ini for the warnings turned off.

Reply:
```
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: df = pd.DataFrame(np.arange(25).reshape(5, 5))

In [4]: grouped = df.groupby(0)

In [5]: grouped.get_group(0).iloc[0, 0]
Out[5]: 0

In [6]: hex(id(df.iloc[0, 0]))
Out[6]: '0x15d491866a8'

In [7]: hex(id(grouped.get_group(0).iloc[0, 0]))
Out[7]: '0x15d491866c0'

In [9]: grouped.get_group(5).iloc[0, 0]
Out[9]: 5

In [10]: hex(id(df.iloc[1, 0]))
Out[10]: '0x15d49186750'

In [11]: hex(id(grouped.get_group(5).iloc[0, 0]))
Out[11]: '0x15d49186768'
```
The ``hex`` should be the memory address and it is different for the same number within the dataframe and the groupby. But, I am a little bit concerned whether this check is valid because the hex seem very similar.

Reply:
This code will change; the tests show the current version works as it should. No need to spend a whole lot of effort on it now.

Reply:
Okay! So I'll merge!